tests/property/ic_api/canister_version/test/tests.ts

```diff
@@ -0,0 +1,129 @@
+
+                        const actor = await getCanisterActor<Actor>('canister');
+
+                        await checkInitVersion(actor, startingVer);
+                        await checkUpgradeVersion(actor, []);
+                        await checkInspectVersion(actor, startingVer);
+                        await checkQueryAndUpdateVersion(
+                            actor,
+                            startingVer + 1n
+                        );
+
+                        const versionsAtDeploy = Array.from(
+                            { length: timesToDeploy },
+                            (_, index) => BigInt(index * 2) + startingVer
+                        );
+
+                        for (const version of versionsAtDeploy) {
+                            execSync(`dfx deploy canister --upgrade-unchanged`);
+
+                            await checkUpgradeVersion(actor, [version]);
+                            await checkQueryAndUpdateVersion(actor, version);
+                        }
+                    }
+                ),
+                defaultPropTestParams
+            );
+        });
+    };
+}
+
+async function checkInitVersion(
+    actor: Actor,
+    expectedVer: bigint
+): Promise<void> {
+    const initCanisterVersion = await actor.getInitCanisterVersion();
+
+    expect(initCanisterVersion).toHaveLength(1);
+    expect(initCanisterVersion[0]).toBeGreaterThanOrEqual(expectedVer);
+}
+
+async function checkUpgradeVersion(
+    actor: Actor,
+    expectedVersion: [bigint] | []
+): Promise<void> {
+    if (expectedVersion.length === 0) {
+        const postUpgradeCanisterVersion =
+            await actor.getPostUpgradeCanisterVersion();
+        expect(postUpgradeCanisterVersion).toHaveLength(0);
+
+        const preUpgradeCanisterVersion =
+            await actor.getPreUpgradeCanisterVersion();
+        expect(preUpgradeCanisterVersion).toHaveLength(0);
+    } else {
+        const postUpgradeCanisterVersionAfterUpgrade =
+            await actor.getPostUpgradeCanisterVersion();
+        expect(
+            postUpgradeCanisterVersionAfterUpgrade[0]
+        ).toBeGreaterThanOrEqual(expectedVersion[0]);
+
+        const preUpgradeCanisterVersionAfterUpgrade =
+            await actor.getPreUpgradeCanisterVersion();
+        expect(preUpgradeCanisterVersionAfterUpgrade[0]).toBeGreaterThanOrEqual(
+            expectedVersion[0]
+        );
+    }
+}
+
+async function checkQueryAndUpdateVersion(
+    actor: Actor,
+    expectedVersion: bigint
+): Promise<void> {
+    expect(await actor.getQueryCanisterVersion()).toBeGreaterThanOrEqual(
+        expectedVersion
+    );
+    expect(await actor.getUpdateCanisterVersion()).toBeGreaterThanOrEqual(
+        expectedVersion
+    );
+    expect(await actor.getQueryCanisterVersion()).toBeGreaterThanOrEqual(
+        expectedVersion + 1n
+    );
+}
+
+async function checkInspectVersion(
+    actor: Actor,
+    expectedVersion: bigint
```

Comment:
Shouldn't these expected versions be exact?

Reply:
If your understanding of the spec is different I would be happy to update it